x-pack/test/apm_api_integration/tests/transactions/transactions_groups_agg_results.ts

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import expect from '@kbn/expect';
+import url from 'url';
+import { FtrProviderContext } from '../../common/ftr_provider_context';
+import archives from '../../common/fixtures/es_archiver/archives_metadata';
+import { registry } from '../../common/registry';
+
+export default function ApiTest({ getService }: FtrProviderContext) {
+  const supertest = getService('supertest');
+
+  const archiveName = 'apm_8.0.0';
+  const { start, end } = archives[archiveName];
+
+  registry.when(
+    'Transaction groups agg results when data is not loaded',
+    { config: 'basic', archives: [] },
+    () => {
+      it('handles the empty state', async () => {
+        const response = await supertest.get(
+          url.format({
+            pathname: `/api/apm/services/opbeans-java/transactions/groups/agg_results`,
+            query: {
+              start,
+              end,
+              uiFilters: '{}',
+              numBuckets: 20,
+              latencyAggregationType: 'avg',
+              transactionType: 'request',
+              transactionNames: JSON.stringify([
+                'DispatcherServlet#doGet',
+                'APIRestController#customers',
+              ]),
+            },
+          })
+        );
+
+        expect(response.status).to.be(200);
+        expect(response.body).to.empty();
+      });
+    }
+  );
+
+  registry.when(
+    'Transaction groups agg results when data is loaded',
+    { config: 'basic', archives: [archiveName] },
+    () => {
+      it('returns the correct data', async () => {
+        const response = await supertest.get(
+          url.format({
+            pathname: `/api/apm/services/opbeans-java/transactions/groups/agg_results`,
+            query: {
+              start,
+              end,
+              uiFilters: '{}',
+              numBuckets: 20,
+              transactionType: 'request',
+              latencyAggregationType: 'avg',
+              transactionNames: JSON.stringify([
+                'DispatcherServlet#doGet',
+                'APIRestController#customers',
+              ]),
+            },
+          })
+        );
+
+        expect(response.status).to.be(200);
+
+        expectSnapshot(Object.keys(response.body)).toMatchInline(`
```

Comment:
FWICT this only include snapshot tests and status checks, but I think we should also add logical assertions, e.g. does it return results for all the requested groups, does it return any data for all statistics, etc. Other API tests should have examples that you can maybe use as an inspiration.

Reply:
@dgieselaar I think the current test already does what you suggested:

> does it return results for all the requested groups

https://github.com/cauemarcondes/kibana/blob/apm-transaction-table-api/x-pack/test/apm_api_integration/tests/transactions/transactions_groups_agg_results.ts#L73-L78

> does it return any data for all statistics

Impact data check:
https://github.com/cauemarcondes/kibana/blob/apm-transaction-table-api/x-pack/test/apm_api_integration/tests/transactions/transactions_groups_agg_results.ts#L80-L86

Latency data check:
https://github.com/cauemarcondes/kibana/blob/apm-transaction-table-api/x-pack/test/apm_api_integration/tests/transactions/transactions_groups_agg_results.ts#L92

Throughput data check:
https://github.com/cauemarcondes/kibana/blob/apm-transaction-table-api/x-pack/test/apm_api_integration/tests/transactions/transactions_groups_agg_results.ts#L133

Error rate data check:
https://github.com/cauemarcondes/kibana/blob/apm-transaction-table-api/x-pack/test/apm_api_integration/tests/transactions/transactions_groups_agg_results.ts#L262

Reply:
These are snapshot tests. Not logical tests. What happens when the archive is updated and this API call returns no data?

Reply:
The API will fail, but if you set --updateSnapshot, the snapshot will be automatically updated. Which IMHO is the same as having a test that does:

`expect(item.latency.length).to.be(10);`

The difference is that I'd have to go manually there and update the new length of the array.

Reply:
I don't mean value tests either. What I mean is something like:

```
expect(body.items.length).to.be(transactionNames.length).
```

or
```
expect(body.items[0].latency).to.be.greaterThan(0);
```

Reply:
To me having a check that only checks the length of an array and not the result of it is kind of useless. If I mess the way the latency is calculated for instance this test won't fail.

But anyway, if that's the way we agreed on testing the API results I'm fine with it.

Reply:
It's just one example. But the point is that it is a logical assertion that answer the question: if I request these transactionNames from this API, does it return data for all transactionNames? Or, does it return a non-null value for all transaction groups (like it should)?

Suppose that somebody else, who did not develop this feature, needs to update the archive. If we have logical assertions next to snapshot or value tests, that process becomes less error-prone. Or, somebody is making changes to this API and updates the snapshots in one go. The chance of a bug not being caught is significant, as snapshot tests are noisey. Logical tests are a safeguard that I feel is important.

Reply:
https://github.com/elastic/kibana/pull/88946/commits/fa2ad695adc3a62f2c0ef761af0166cf1de0ab2f#diff-3580d90605bfa0ab31989f242d5f937e9e2d812f9e914dbbb4cf3b46ba5cdb36R68